Extend the  ZKProofs module (#875)
Polkadot-Blockchain-Academy/pba-content: syllabus/1-Cryptography/zk-proofs/slides.md

---
title: ZK proofs
description: Introduction to zero-knowledge proofs and zk-SNARKS
duration: 3.5 hour
---

# ZK Proofs

---

# Outline

<pba-flex center>

1. [ZK Proofs overview](#zk-proofs)
2. [Examples](#simple-zk-example)
3. [zk-SNARKs](#zk-snarks)
4. [Applications of zk-SNARKs](#zk-application)
5. [Under the hood of zk-SNARKs](#zk-practice)
6. [Generating ZK-proofs using Circom and snarkjs](#circom-snarkjs)

</pba-flex>

---

## ZK Proofs

How do we do private operations on a public blockchain<br />and have everyone know that they were done correctly?

Notes:

(we are working on substrate support for these and will use them for protocols)

---

## What is a ZK Proof?

- A prover wants to convince a verifier that something is true without revealing why it is true.

> - They are mostly interactive protocols, but mostly we'll turn them  non-interactive and deal with their non-interactive variety. 

Reviewer: Students don't know the definitions of interactive/non-interactive at this point, so probably add a speaker note to explain the distinction briefly before you get to the proper definition later on.

> - They are mostly interactive protocols, but mostly we'll turn them  non-interactive and deal with their non-interactive variety. 

Reviewer 2: ```suggestion
- They are mostly interactive protocols, but mostly we'll turn them non-interactive and deal with their non-interactive variety. 
```

---

## What can we show?

- NP relation: `function(statement, witness) -> bool`

- Prover knows a witness for a statement:

  - They want to show that they know it (_a proof of knowledge_)

  - ... Without revealing anything about the witness (_ZK_)

---

## ZK Proof Interface

- NP relation: `function(statement, witness) -> bool`

- `prove(statement, witness) -> proof`

- `verify(statement, proof) -> bool`

---

## Examples
---

## ZK Proof Example

_Example:_ Schnorr signatures are ZK Proofs
- $(s, e)$ such that $s = (\textrm{Random blinding factor }r) - \\ \textrm{Private Key} \times e$

> - They show that the prover knows the private key (the discrete log of the public key) without revealing anything about it.

Reviewer 2: I think you should tread lightly on this. Even mentioning discrete log I dont even think is necessary. I think it can just be generalized to the prover contains knowledge of a secret key and can prove it to the verifier without revealing it.

Something to this affect but if you do decide to continue just remember to tread lightly.

- The statement is the public key and the witness the private key.

---

## The ZK Proof Example we study in this module

_Example:_ Prover knows a non-trivial factorization of N.

- Prover wants to prove that they know $N = r \times s$  without revealing $r$ or $s$.

- The Prover should convince us they know the two non-trivial integer $r$ and $s$ that is $r,s \neq 1$ such that:

 $r\times s = N$ 

- Without revealing any other information about $r$ or $s$.

> - The statement is $N$ and the witnesses are $r$ and $s$.

Reviewer 2: In this example I think the math is appropriate because you are giving a concrete example using factorization.

---
## ZK Proof properties.
 - Completeness: If the claim is true, then it *must* pass `verify(statement, proof) == true`
 - Statistical Knowledge Soundness: If the prover does not know the witness then `verify(statement, proof) == false` with high probability.
 - Zero Knowledge: the proof reveals nothing about the witness that was not revealed by the statement itself.
 - The common way of implementing zero knowledge protocol is by means of zk-SNARK.

Notes:
 
 - We prove Zero knowldege by showing a party without any knowledge of the witness, could fool an external party who was not been part of the interaction that they interacted with the prover and got a valid proof form the prover. 
---

## zk-SNARKs

---

## zk-SNARK

**Z**ero-**K**nowledge **S**uccinct **N**on-interactive **Ar**gument of **K**nowledge

- **Zero knowledge** - the proof reveals nothing about the witness that was not revealed by the statement itself.
- **Succinct** - the proof is small

> - **Non-interactie** - Does not  require live interaction between prover and the verifier.

Reviewer: ```suggestion
- **Non-interactive** - Does not require live interaction between the prover and verifier.
```

- **Proof of knowledge** - if you can compute correct proofs of a statement, you should be able to compute a witness for it.

---

## What can we show?

- NP relation: `function(statement, witness) -> bool`

  - They want to show that they know it (_a proof of knowledge_)

  - ... Without revealing anything about the witness (_ZK_)

- With a small proof even if the witness is large (_succinctness_)

---

## What can we show?

> - There are many SNARK schemes to produce succinct ZK proofs of knowledge (_ZK-SNARKs_) for every NP relation. We concentrate on PLONK in this course.

Reviewer 2: Just be ready to explain NP in a succinct way :)

No pun intended there.

Author: I thnk all mentions of NP are inherited from @AlistairStewart 's slide. I could remove them if there is a consensus on that.

Reviewer 2: I think it is fine to keep in. I am just leaving a note that you should not assume everyone is familiar is all.

---

## Applications of zk-SNARKs

---

## ZK Proof Scaling

- A small amount of data, a ZK proof, and execution time can be used to show properties of a much larger dataset which the verifier doesn't need to know. 

- "doesn't want to" know as opposed to "is not supposed to" know.

---

## Scaling via ZK Proofs in Blockchain

- Large amount of data - a blockchain
- Verifier is e.g. an app on a mobile phone

Notes:

e.g. Mina do a blockchain with a constant size proof (of correctness of execution and consensus) using recursive SNARKs.

---

## Scaling via ZK Proofs in Blockchain

- The verifier is a blockchain: very expensive data and computation costs.

- Layer 2s using ZK rollups

Notes:

Of which Ethereum has many, ZKsync, ZKEVM etc.
Polkadot already scales better!

> ---

Reviewer 2: One of the original requirements of this module was to have no mention of blockchains. Just an FYI. We are building up to the blockchain datastructure. Going into Layer 2's is too premature.

Author: I do not believe I have added any mention of blockchain, they are all from previous slides.

Reviewer 2: Up to you. I am just giving you what was told to us.

## Privacy

<pba-flex center>

A user has private data, but we can show<br />publicly that this private data is correctly used.<br />
An example would a private cryptocurrency:

- Keep who pays who secret

> - Keep amounts secret, <br /> _But show they are positive!_

Reviewer 2: This I think is ok though. It is basic enough at this point. We do mention Bitcoin in some of the early hashing lectures as well. I remember Gav doing so aswell.

</pba-flex>

Notes:

You can do some of keeping amounts secret without ZK-SNARKs, but the positive part is difficult.
To do everything well, ZK-SNARKs are needed in e.g. ZCash and its many derivatives e.g. Manta.

---

## Under the hood of zk-SNARKs

Notes:

 - We are going to talk about the elementary math behind zk-SNARKs.
 - The goal is to familiarize you with the magic behind the zk-SNARKs.
 - Don't worry if you are not able to follow. Not essential.
---

## Making a SNARK out of the factorization problem
- The trick  is to transform our problem of proving the knowledge of factors (witnesses) into:

> - A problem of knowledge of a certain polynomials.

Reviewer: ```suggestion
- A problem of knowledge of certain polynomials.
```

- Then verifier could ask me questions about those polynomials, and if the prover answers correctly,

> - The verifier could be fairly confidant that the prover knows that polynomial hence also the witness.

Reviewer: ```suggestion
- The verifier could be fairly confident that the prover knows that polynomial hence also the witness.
```

---

## Making an SNARK for knowledge of factors problem
- A routine way of to turning the problem into a polynomial is:
1. To represents our problem into an arithmetic circuit.

> 2. Then are algorithms such as PLONK for representing the circuit as few univariate polynomials.

Reviewer: There are? Then use? Not sure what this was supposed to be

Notes: 
 - Mathematically circuit is a n-variate polynomials, with some of the variables are public and some are not.
---

## The conditions the prover's solution must satisfy
- $r \times s = N$
- We also need to make sure that prover doesn't fool us with trivial factors.
- To prevent them from using $r = 1$ or $s = 1$ we ask them to invert $r - 1$ and $s-1$:
  - $(r-1)*\frac{1}{r - 1} = 1$
  - $(s-1)*\frac{1}{s - 1} = 1$

---

## Overflow prevention
- Our polynomials are all defined modulo some prime $p$
- We need to prevent the prover from fooling us with a factorization like 

    $r \times s = N + q\times p$ where $q \neq 0$.

> - This happens if $r$ and $s$ are two big and $r \times s$ overflows over $p$

Reviewer: ```suggestion
- This happens if $r$ and $s$ are too big and $r \times s$ overflows over $p$
```

- We should make sure that $r$ and $s$ are small.
- We use binary decomposition for that.
- $r = r_{0} + 2r_{1} + 4r_{2}$ where 

> - We should only allow them use 0 or1 for $r_{i}$'s so they need to satisfies:

Reviewer: ```suggestion
- We should only allow them to use 0 or 1 for $r_{i}$'s so they need to satisfy:
```

- ... without knowing the polynomial: This is done using *polynomial commitment*.

---

## Universal PLONK Gate

- Supppose we have a left input $a$ and a right input $b$ and we are doing some addition and multiplication with them and the output is $c$. Then we could encode all of these operations as:

$Q_l\times a + Q_r \times b + Q_o \times c + Q_m \times a\times b + Q_c = 0$

- for some constant $Q_l$ $Q_r$ $Q_o$ $Q_m$ and $Q_c$
- in fact all the operation we discussed can be written using one of these gates.

---

## Gate table for factorization
 $r \times s = N$

 $Q_l\times a + Q_r\times b + Q_o\times c + Q_m\times a\times b + Q_c = 0$

 
 <img style="height: 200px; padding-left:100px" src="./img/gate-table-factorization.png" />

---

## Gate table for left input to be small and not 1
  
 $r_{01} = r_{0} + 2r_{1}$
 
 $r = r_{01} + 4r_{2}$ 

 $r_{i} \times (r_{i} - 1) = 0 \Rightarrow r_{i}^2 - r_{i} = 0$
 
 $(r-1)*\frac{1}{r - 1} = 1 \Rightarrow r\frac{1}{r - 1} - \frac{1}{r - 1} = 1$
 
 $Q_l\times a + Q_r\times b + Q_o\times c + Q_m\times a\times b + Q_c = 0$ 

<img style="height: 500px; padding-left:100px" src="./img/gate-table-left-input-less-than-8-and-not-1.png" />

---

> ## Gate table for right input to be integer and not 1

Reviewer 2: ```suggestion
## Gate table for the right input to be an integer and not 1
```
